Objective-C/Tests/TLSIdentityTest.m

```diff
@@ -0,0 +1,339 @@
+
+@end
+
+#define kServerCertLabel @"CBL-Server-Cert"
+#define kClientCertLabel @"CBL-Client-Cert"
+
+@implementation TLSIdentityTest
+
+- (CFTypeRef) findInKeyChain: (NSDictionary*)params {
+    CFTypeRef result = NULL;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)params, &result);
+    if (status != errSecSuccess) {
+        if (status == errSecItemNotFound)
+            return nil;
+        else
+            NSLog(@"Couldn't get an item from the Keychain: %@ (error: %d)", params, status);
+    }
+    Assert(result != nil);
+    return result;
+}
+
+- (NSData*) publicKeyHashFromCert: (SecCertificateRef)certRef {
+    // Get public key from the certificate:
+    SecKeyRef publicKeyRef;
+    if (@available(iOS 12, macOS 10.14, *)) {
+        publicKeyRef = SecCertificateCopyKey(certRef);
+    } else {
+    #if TARGET_OS_IOS
+        if (@available(iOS 10.3, *))
+            publicKeyRef = SecCertificateCopyPublicKey(certRef);
+        else
+            Assert(false, @"Not Supported by this OS version");
+    #else
+        OSStatus status = SecCertificateCopyPublicKey(certRef, &publicKeyRef);
+        Assert(status == errSecSuccess);
+    #endif
+    }
+    Assert(publicKeyRef);
+                
+    if (@available(iOS 10.0, macOS 10.12, *)) {
+        NSDictionary* attrs = CFBridgingRelease(SecKeyCopyAttributes(publicKeyRef));
+        return [attrs objectForKey: (id)kSecAttrApplicationLabel];
+    } else {
+        Assert(false, @"Not Supported by this OS version");
+    }
+    return nil;
+}
+
+- (void) checkIdentityInKeyChain: (CBLTLSIdentity*)identity {
+    NSArray* certs = identity.certs;
+    Assert(certs.count > 0);
+    
+    // Check Certificate:
+    SecCertificateRef certRef = (__bridge SecCertificateRef)certs[0];
+
+    // Check Private Key:
+    NSData* publicKeyHash = [self publicKeyHashFromCert: certRef];
+    SecKeyRef privateKeyRef = (SecKeyRef)[self findInKeyChain: @{
+        (id)kSecClass:                  (id)kSecClassKey,
+        (id)kSecAttrKeyType:            (id)kSecAttrKeyTypeRSA,
+        (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
+        (id)kSecAttrApplicationLabel:   publicKeyHash,
+        (id)kSecReturnRef:              @YES
+    }];
+    Assert(privateKeyRef != nil);
+    
+    // Get public key from Cert via Trust:
+    SecTrustRef trustRef;
+    SecTrustCreateWithCertificates(certRef, SecPolicyCreateBasicX509(), &trustRef);
+    SecKeyRef publicKeyRef = SecTrustCopyPublicKey(trustRef);
+    Assert(publicKeyRef != nil);
+    
+    CFErrorRef errRef = nil;
+    NSData* publicKeyData = CFBridgingRelease(SecKeyCopyExternalRepresentation(publicKeyRef, &errRef));
+    Assert(errRef == nil);
+    AssertNotNil(publicKeyData);
+    Assert(publicKeyData.length > 0);
+    NSLog(@"%lu", (unsigned long)publicKeyData.length);
+}
+
+- (void) deleteAllItemsInKeyChain {
+    NSArray *classes = @[(id)kSecClassKey, (id)kSecClassCertificate, (id)kSecClassIdentity];
```

Comment:
is this going to delete all certificates and identities form keychain?

Reply:
It's fine for iOS as the keychain items are app's bound. For macOS, there could be some side effects. Will try to improve this.

Reply:
Actually that method is not used anymore. Forgot that I already changed to delete the identity directly which is definitely safer.